server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java

```diff
@@ -129,28 +133,13 @@ public static TranslogWriter create(ShardId shardId, String translogUUID, long f
         }
     }
 
-    /**
-     * If this {@code TranslogWriter} was closed as a side-effect of a tragic exception,
-     * e.g. disk full while flushing a new segment, this returns the root cause exception.
-     * Otherwise (no tragic exception has occurred) it returns null.
-     */
-    public Exception getTragicException() {
-        return tragedy;
-    }
-
     private synchronized void closeWithTragicEvent(final Exception ex) {
-        assert ex != null;
-        if (tragedy == null) {
-            tragedy = ex;
-        } else if (tragedy != ex) {
-            // it should be safe to call closeWithTragicEvents on multiple layers without
-            // worrying about self suppression.
-            tragedy.addSuppressed(ex);
-        }
-        try {
-            close();
-        } catch (final IOException | RuntimeException e) {
-            ex.addSuppressed(e);
+        if (tragedy.setTragicException(ex)) {
```

Comment:
why change the semantics here to only call close when setting the tragedy the first time? Let's keep the existing semantics and make `setTragicException` return void

Reply:
@ywelsch what is the reason to call close twice if every time we set tragic exception close is called?